### PR TITLE
Repair cluster storage calculation method(#66789)

### PR DESCRIPTION
**Pull request by calm4wei**
Complete calculation of cluster storage capacity,
1. Deduplication based on ip address;
2. Single machine multi-instance accumulation based on mount point;
3. Data disk deduplication based on the mount disk;

### CODE REVIEW

**Comment by nik9000** [comment]
I'll leave this for someone who knows that nodes stats stuff better than I do, but I imagine this'll want a test update so we don't accidentally break it later.

**Comment by calm4wei** [comment]
> I'll leave this for someone who knows that nodes stats stuff better than I do, but I imagine this'll want a test update so we don't accidentally break it later.

Okay, I will add a test case to this.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodes.java [hunk 1]
```diff
@@ -76,24 +77,38 @@
...
-            final InetAddress inetAddress = publishAddress.address().getAddress();
-            if (!seenAddresses.add(inetAddress)) {
-                continue;
-            }
-            if (nodeResponse.nodeStats().getFs() != null) {
-                this.fs.add(nodeResponse.nodeStats().getFs().getTotal());
+                final InetAddress inetAddress = publishAddress.address().getAddress();
+                Iterator<FsInfo.Path> pathIterator = nodeResponse.nodeStats().getFs().iterator();
+                Set<String> fsInfoPaths;
+                if (!inetAddressFsPaths.containsKey(inetAddress)){
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodes.java:97]: another super minor nit
```suggestion
                if (!inetAddressFsPaths.containsKey(inetAddress)) {
```

File: server/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodes.java [hunk 2]
```diff
@@ -76,24 +77,38 @@
...
+                    fsInfoPaths = new HashSet<>();
+                    inetAddressFsPaths.put(inetAddress, fsInfoPaths);
+                } else {
+                    fsInfoPaths = inetAddressFsPaths.get(inetAddress);
+                }
+
+                // single machine with multiple instances, accumulating different mount data directories
+                while (pathIterator.hasNext()) {
+                    FsInfo.Path path = pathIterator.next();
+                    if (!fsInfoPaths.contains(path.getMount())){
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodes.java:107]:
```suggestion
                    if (!fsInfoPaths.contains(path.getMount())) {
```

File: server/src/test/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodesTests.java [hunk 3]
```diff
@@ -129,4 +174,37 @@ private static NodeInfo createNodeInfo(String nodeId, String transportType, Stri
...
+    private static ClusterStatsNodeResponse createNodeStatResponse(String nodeId, String hostname, Integer port, FsInfo.Path[] fsInfoPaths) throws Exception {
+
+        DiscoveryNode discoveryNode = new DiscoveryNode(nodeId, new TransportAddress(new InetSocketAddress(hostname, port)), null);
+        return new ClusterStatsNodeResponse(new DiscoveryNode(nodeId, new TransportAddress(new InetSocketAddress(hostname, port)), null), null,
+            createNodeInfo2(nodeId, hostname, port),
+            createNodeStats(discoveryNode, fsInfoPaths),
+            null);
+    }
+
+    private static NodeInfo createNodeInfo2(String nodeId, String hostname, Integer port){
```
Comment (dakrone) [inline server/src/test/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodesTests.java:187]:
```suggestion
    private static NodeInfo createNodeInfo2(String nodeId, String hostname, Integer port) {
```

File: server/src/test/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodesTests.java [hunk 4]
```diff
@@ -129,4 +174,37 @@ private static NodeInfo createNodeInfo(String nodeId, String transportType, Stri
...
+
+    private static NodeStats createNodeStats(DiscoveryNode discoveryNode, FsInfo.Path[] fsInfoPaths) throws UnknownHostException {
+
+        long timestamp = 1608809284528l;
+        return new NodeStats(discoveryNode, timestamp, null, null ,null, null,
+            null, new FsInfo(timestamp, null, fsInfoPaths),
+            null, null, null, null,null, null, null, null);
+    }
+
+    private static FsInfo.Path createFsInfoPath(String path, String mount, long total, long free, long available){
```
Comment (dakrone) [inline server/src/test/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodesTests.java:207]:
```suggestion
    private static FsInfo.Path createFsInfoPath(String path, String mount, long total, long free, long available) {
```

**Review by dakrone — APPROVED** [review]
@calm4wei thanks for opening this, I left very very minor nits, but otherwise this LGTM.

**Comment by dakrone** [comment]
@elasticmachine update branch

**Comment by dakrone** [comment]
@elasticmachine test this please

**Comment by dakrone** [comment]
@calm4wei looks like this causes a couple of test failures, one related to checkstyle, another to an NPE in the stats now. Can you please check out the failures and resolve them?

**Comment by dakrone** [comment]
Hi @calm4wei, we haven't heard back from you in a while, so I'm going to close this for now. Feel free to let us know if you're still interested in working on it and we can re-open this, thanks!

**Comment by calm4wei** [comment]
> Hi @calm4wei, we haven't heard back from you in a while, so I'm going to close this for now. Feel free to let us know if you're still interested in working on it and we can re-open this, thanks!

Sorry, I did not continue to pay attention to this issue some time ago. Now I want to continue to repair it, can you open this pr?

**Pull request closed without merging**